pkg/watch/health.go

```diff
@@ -0,0 +1,240 @@
+
+// Contains method for watching the consul reality store to
+// track services running on a node. A manager method:
+// WatchHealth tracks the reality store and manages
+// a health checking go routine for each service in the
+// reality store
+
+// PodWatch houses a pod's manifest, a channel to kill the
+// pod's goroutine if the pod is removed from the reality
+// tree, and a bool that indicates whether or not the pod
+// has a running MonitorHealth go routine
+type PodWatch struct {
+	manifest   kp.ManifestResult
+	shutdownCh chan bool
+	hasMonitor bool // indicates whether this pod is being monitored
+	logger     *logging.Logger
+}
+
+// WatchHealth is meant to be a long running go routine.
+// WatchHealth reads from a consul store to determine which
+// services should be running on the host. WatchHealth
+// runs a CheckHealth routine to monitor the health of each
+// service and kills routines for services that should no
+// longer be running.
+func WatchHealth(node, consul, authtoken string, logger *logging.Logger, shutdownCh chan struct{}) {
+	var store kp.Store
+	tochan := make(chan bool)
+	pods := []*PodWatch{}
+	if authtoken != "" {
+		store = kp.NewConsulStore(kp.Options{
+			Address: consul,
+			HTTPS:   false,
+			Token:   authtoken,
+			Client:  net.NewHeaderClient(nil, http.DefaultTransport),
+		})
+	} else {
+		store = kp.NewConsulStore(kp.Options{
+			Address: consul,
+			HTTPS:   false,
+			Client:  net.NewHeaderClient(nil, http.DefaultTransport),
+		})
+	}
+
+	go startTimer(tochan, POLL_KV_FOR_PODS)
+	for {
+		select {
+		case _ = <-tochan:
+			// start timer again
+			go startTimer(tochan, POLL_KV_FOR_PODS)
+
+			// check if pods have been added or removed
+			// starts monitor routine for new pods
+			// kills monitor routine for removed pods
+			pods = updateHealthMonitors(store, pods, node, logger)
+		case _ = <-shutdownCh:
+			return
+		}
+	}
+}
+
+// Monitor Health is a go routine that runs as long as the
+// service it is monitoring. Every HEALTHCHECK_INTERVAL it
+// performs a health check and writes that information to
+// consul
+func (p *PodWatch) MonitorHealth(node string, store kp.Store, shutdownCh chan bool) {
+	statusCheck := fmt.Sprintf(kp.GetStatusCheck(), p.manifest.Manifest.StatusPort)
+	tochan := make(chan bool)
+
+	go startTimer(tochan, HEALTHCHECK_INTERVAL)
+	for {
+		select {
+		case _ = <-tochan:
+			p.checkHealth(statusCheck, node, store)
+			go startTimer(tochan, HEALTHCHECK_INTERVAL)
+		case _ = <-shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *PodWatch) checkHealth(healthCheck, node string, store kp.Store) {
+	healthstate, res, err := check(healthCheck)
+	health := health.Result{
+		ID:      p.manifest.Manifest.Id,
+		Node:    node,
+		Service: p.manifest.Manifest.Id,
+		Status:  healthstate,
+		Output:  res,
+	}
+
+	err = writeToConsul(health, store)
+	if err != nil {
+		p.logger.WithField("CONSUL WRITE ERROR", err).Fatalln("failed to write health data to consul")
```

Comment:
The "WithField" is for a key in a JSON document most likely. You probably just want to use the string "err" instead of "CONSUL WRITE ERROR" here.